### PR TITLE
[Workspace] Add workspace column to saved objects page

### LINKED ISSUE #6098: [Workspace] Add workspace column and filter to saved objects page
**Is your feature request related to a problem? Please describe.**

When workspace feature turned on, saved objects will have a new attribute named workspace, to better filter and show which workspace they belong to, we are going to add a new workspace column in saved objects page.

A mockup page

image

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**
#4944

### PR DESCRIPTION
**Pull request by Hailong-am**
### Description

This PR will add workspaces column to saved objects page when workspace feature enabled, the workspaces column will show the workspace names which the saved object belongs to.

### Issues Resolved

partial of #6098  #4944

## Screenshot

image

## Testing the changes

1. Start OpenSearch Dashboards with workspace feature enabled `yarn start --workspace.enabled=true`
2. Create test workspace and import sample data
```shell
curl --request POST \
  --url http://localhost:5601/api/workspaces \
  --header 'Content-Type: application/json' \
  --header 'osd-xsrf: 1' \
  --data '{
 "attributes": {
  "name": "my-test-workspace",
  "features": ["*"]
 }
}'

# response
{
 "success": true,
 "result": {
  "id": "yRJbeT"
 }
}
```

Add existing saved objects to workspace just created
```
POST .kibana/_update_by_query
{
  "query": {
    "term": {
      "type": {
        "value": "dashboard"
      }
    }
  },
  "script": {
    "source": """
    ctx._source.workspaces = ['yRJbeT'];
    """,
    "lang": "painless"
  }
}
```

3. Open saved objects page http://localhost:5601/app/management/opensearch-dashboards/objects to see the result

### Check List

- [x] All tests pass
  - [x] `yarn test:jest`
  - [x] `yarn test:jest_integration`
- [x] New functionality includes testing.
- [ ] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/table.tsx [hunk 1]
```diff
@@ -231,9 +238,19 @@ export class Table extends PureComponent<TableProps, TableState> {
...
           }
-          return (
-            <EuiLink href={basePath.prepend(path)}>{title || getDefaultTitle(object)}</EuiLink>
-          );
+          let inAppUrl = basePath.prepend(path);
+          if (object.workspaces?.length) {
+            if (currentWorkspaceId) {
+              inAppUrl = formatUrlWithWorkspaceId(path, currentWorkspaceId, basePath);
+            } else {
+              // first workspace user have permission
```
Comment (ruanyl) [inline src/plugins/saved_objects_management/public/management_section/objects_table/components/table.tsx:246]: `first` -> `filter`?

Comment (ruanyl) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/table.tsx:246]: Oh I see the point, it just need to find the first workspace that matches, maybe can simply use `object.workspaces.find`

Comment (Hailong-am) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/table.tsx]: Thanks, updated to `find` method.

Comment (ruanyl) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/table.tsx:246]: Still can see `object.workspaces.filter`, am I looking at the right version?

Comment (Hailong-am) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/table.tsx:246]: ah, seem have code merge issue when rebase from main. let me redo the change.

**Comment by seraphjiang** [comment]
@Hailong-am  @ruanyl  @kgcreative @zengyan-amazon 

Got some questions

1. Is it the plan to support make one saved object could be belong to multiple workspace. If so, it make sense we see one object could have multiple workspaces information. 
2. Another question, in the left nav menu, we plan to let cx select a workspace right, suppose that will set current workspace context to all content/page on right side. do we still need to show workspace info in each page?

image

image

**Comment by Hailong-am** [comment]
> * Is it the plan to support make one saved object could be belong to multiple workspace. If so, it make sense we see one object could have multiple workspaces information.

1. share saved object to different workspace is not in MVP scope, but we might support this in future.
2. when cx in `home` workspace means not in any workspace, cx could see all saved objects which cx have access across different workspaces, to have this column is helpful.

**Comment by Flyingliuhub** [comment]
@Hailong-am Did you have a chance to add the feature flag `data_source.enabled: false` when you were testing? Workspaces should work with multiple data sources in the near future, so please make sure that the workspace functions properly with the multiple data source feature.

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/table.test.tsx [hunk 2]
```diff
@@ -115,6 +116,50 @@ describe('Table', () => {
...
+    const props = {
+      ...defaultProps,
+      availableWorkspaces: [{ id: 'ws-1', name: 'My workspace' } as WorkspaceAttribute],
+      items: [item],
+    };
+    // not in a workspace
+    let component = shallowWithI18nProvider(<Table {...props} />);
+
+    let table = component.find('EuiBasicTable');
+    let columns = table.prop('columns') as any[];
```
Comment (Flyingliuhub) [inline src/plugins/saved_objects_management/public/management_section/objects_table/components/table.test.tsx:145]: `any[]` do we have more specific types here?

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx [hunk 3]
```diff
@@ -172,13 +179,17 @@ export class SavedObjectsTable extends Component<SavedObjectsTableProps, SavedOb
 
   componentDidMount() {
     this._isMounted = true;
+
+    this.subscribleWorkspace();
```
Comment (Flyingliuhub) [inline src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx:183]: Is it typo here? `subscribleWorkspace`, `subscribeWorkspace`?

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx [hunk 4]
```diff
@@ -172,13 +179,17 @@ export class SavedObjectsTable extends Component<SavedObjectsTableProps, SavedOb
...
     this.fetchSavedObjects();
     this.fetchCounts();
   }
 
   componentWillUnmount() {
     this._isMounted = false;
     this.debouncedFetchObjects.cancel();
```
Comment (Flyingliuhub) [inline src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx:190]: Do we need to manually cancel it? is it automatically handles cleanup when the component unmounts?

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx (hunk 3 above)
Comment (Hailong-am) [reply src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx:183]: good catch, fix the typo.

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx (hunk 4 above)
Comment (Hailong-am) [reply src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx:190]: Yes, we do have unsubscribe in `componentWillUnmount` method. Refactor to a `unSubscribeWorkspace` method to make it more clear.

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/table.test.tsx (hunk 2 above)
Comment (Hailong-am) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/table.test.tsx:145]: Yes, updated to specific type `Array<{ render: (id: string, record: unknown) => React.ReactElement }>`

**Comment by Hailong-am** [comment]
> @Hailong-am Did you have a chance to add the feature flag `data_source.enabled: true` when you were testing? Workspaces should work with multiple data sources in the near future, so please make sure that the workspace functions properly with the multiple data source feature.

@Flyingliuhub Good suggestion. Did a quick testing for both workspace and data_source enabled case, data source/index pattern associated with workspace display correctly.

image

**Comment by xinruiba** [comment]
> > @Hailong-am Did you have a chance to add the feature flag `data_source.enabled: true` when you were testing? Workspaces should work with multiple data sources in the near future, so please make sure that the workspace functions properly with the multiple data source feature.
> 
> @Flyingliuhub Good suggestion. Did a quick testing for both workspace and data_source enabled case, data source/index pattern associated with workspace display correctly.
> 
> image

Thanks for the testing @Hailong-am. Would be good to also test MD scenarios~

1. Create a remote dataSource
2. Add sample data
3. Check dashboard
4. Remove sample data
5. Update remote dataSource.

Thanks~

**Comment by Hailong-am** [comment]
> Thanks for the testing @Hailong-am. Would be good to also test MD scenarios~
> 
> 1. Create a remote dataSource
> 2. Add sample data
> 3. Check dashboard
> 4. Remove sample data
> 5. Update remote dataSource.
> 
> Thanks~

@xinruiba upload the video about the test steps 
https://github.com/opensearch-project/OpenSearch-Dashboards/assets/113890546/ade45329-c713-4a68-976c-04cb3a8ddf2e

**Comment by xinruiba** [comment]
> > Thanks for the testing @Hailong-am. Would be good to also test MD scenarios~
> > 
> > 1. Create a remote dataSource
> > 2. Add sample data
> > 3. Check dashboard
> > 4. Remove sample data
> > 5. Update remote dataSource.
> > 
> > Thanks~
> 
> @xinruiba upload the video about the test steps https://github.com/opensearch-project/OpenSearch-Dashboards/assets/113890546/ade45329-c713-4a68-976c-04cb3a8ddf2e

Appreciate, looks good to me~

**Review by ruanyl — APPROVED** [review]

**Review by Flyingliuhub — APPROVED** [review]

**Pull request merged**